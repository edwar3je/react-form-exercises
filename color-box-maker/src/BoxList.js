import { useState } from 'react';
import { v4 as uuidv4} from 'uuid';
import './BoxList.css';
import Box from './Box';
import NewBoxForm from './NewBoxForm';

const BoxList = () => {
    const initialState = [{
        height: "50px",
        width: "50px",
        backgroundColor: "gold",
        id: uuidv4()
    }];
    
    const [boxes, setBoxes] = useState(initialState);

    const addBox = (height, width, backgroundColor, id=uuidv4()) => {
        setBoxes(boxes => [...boxes, {height, width, backgroundColor, id}]);
    }

    const removeBox = (uniqueId) => {
        setBoxes(
            boxes => boxes.filter(
                ({ id }) => id !== uniqueId
            )
        )
    };

    return (
        <div className="box-list-container">
            <NewBoxForm addBox={addBox}/>
            <div className="boxes-container">
                {boxes.map(({height, width, backgroundColor, id}) => <Box height={height} width={width} backgroundColor={backgroundColor} remove={removeBox} id={id} key={id} />)}
            </div>
        </div>
    );
}

export default BoxList;