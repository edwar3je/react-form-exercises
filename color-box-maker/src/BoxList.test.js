import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import BoxList from './BoxList';

it("renders without crashing", () => {
    render(<BoxList />);
});

it("matches the snapshot", () => {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it("displays initial data from useState", () => {
    const { container } = render(<BoxList />);
    const firstSquare = container.querySelector('div#gold-50px-50px');
    expect(firstSquare).toBeInTheDocument();
});

it("adds a square to the DOM", () => {
    const { container, queryByText, queryByLabelText } = render(<BoxList />);

    // ensure the square we want to add isn't in the DOM already
    expect(container.querySelector('div#pink-25px-30px')).not.toBeInTheDocument();

    // extract the DOM values and use fireEvent to add new values to the input
    const inputColor = queryByLabelText("Background Color");
    const inputHeight = queryByLabelText("Height");
    const inputWidth = queryByLabelText("Width");
    const submitButton = queryByText("Add New Box");
    fireEvent.change(inputColor, {target: {value: 'pink'}});
    fireEvent.change(inputHeight, {target: {value: '25px'}});
    fireEvent.change(inputWidth, {target: {value: '30px'}});
    fireEvent.click(submitButton);

    // ensure the new DOM element exists after form submission
    expect(container.querySelector('div#pink-25px-30px')).toBeInTheDocument();
});

it("makes input values blank strings after adding a square to the DOM", () => {
    const { queryByText, queryByLabelText } = render(<BoxList />);

    // extract the DOM values and see if the values are currently blank
    const inputColor = queryByLabelText("Background Color");
    const inputHeight = queryByLabelText("Height");
    const inputWidth = queryByLabelText("Width");
    const submitButton = queryByText("Add New Box");
    expect(inputColor.value).toBe("");
    expect(inputHeight.value).toBe("");
    expect(inputWidth.value).toBe("");

    // add values, test to ensure they're there and fire event to submit form
    fireEvent.change(inputColor, {target: {value: 'pink'}});
    fireEvent.change(inputHeight, {target: {value: '25px'}});
    fireEvent.change(inputWidth, {target: {value: '30px'}});
    expect(inputColor.value).toBe("pink");
    expect(inputHeight.value).toBe("25px");
    expect(inputWidth.value).toBe("30px");
    fireEvent.click(submitButton);

    // see if the values for inputColor, inputHeight and inputWidth are blank
    expect(inputColor.value).toBe("");
    expect(inputHeight.value).toBe("");
    expect(inputWidth.value).toBe("");
});

it("removes a square from the DOM", () => {
    const { queryByText, queryByLabelText, container } = render(<BoxList />);
    
    // demonstrate that the first square and remove button are on the document
    const firstSquare = container.querySelector('div#gold-50px-50px');
    const firstRemoveButton = container.querySelector('button#gold-50px-50px')
    expect(firstSquare).toBeInTheDocument();
    expect(firstRemoveButton).toBeInTheDocument();

    // extract DOM values (inputs/button) and fireEvent to add new square
    const inputColor = queryByLabelText("Background Color");
    const inputHeight = queryByLabelText("Height");
    const inputWidth = queryByLabelText("Width");
    const submitButton = queryByText("Add New Box");
    fireEvent.change(inputColor, {target: {value: 'pink'}});
    fireEvent.change(inputHeight, {target: {value: '25px'}});
    fireEvent.change(inputWidth, {target: {value: '30px'}});
    fireEvent.click(submitButton);

    // demonstrate that the second square and remove button are on the document
    const secondSquare = container.querySelector('div#pink-25px-30px');
    const secondRemoveButton = container.querySelector('button#pink-25px-30px')
    expect(secondSquare).toBeInTheDocument();
    expect(secondRemoveButton).toBeInTheDocument();

    // click on first remove button and show that the first square is not on the document (but the second square is)
    fireEvent.click(firstRemoveButton);
    expect(firstSquare).not.toBeInTheDocument();
    expect(secondSquare).toBeInTheDocument();

    // click on the second remove button and show that the second square is not on the document
    fireEvent.click(secondRemoveButton);
    expect(secondSquare).not.toBeInTheDocument();
});