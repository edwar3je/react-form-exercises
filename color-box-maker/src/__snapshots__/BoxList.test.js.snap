// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="box-list-container"
  >
    <form>
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        placeholder="width"
        type="text"
        value=""
      />
      <p />
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        name="height"
        placeholder="height"
        type="text"
        value=""
      />
      <p />
      <label
        for="backgroundColor"
      >
        Background Color
      </label>
      <input
        class="backgroundColor"
        id="backgroundColor"
        name="backgroundColor"
        placeholder="background color"
        type="text"
        value=""
      />
      <p />
      <button
        type="submit"
      >
        Add New Box
      </button>
    </form>
    <div
      class="boxes-container"
    >
      <div
        class="color-box"
        id="gold-50px-50px"
        style="height: 50px; width: 50px; background-color: gold;"
      />
      <button
        class="remove-button"
        id="gold-50px-50px"
      >
        X
      </button>
    </div>
  </div>
</DocumentFragment>
`;
